### ISSUE TITLE
Does there exist a parallel version of this algorithm for millions of items like on SPARK?

### ISSUE THREAD

**Issue by cmxcn** [hasteck/MRF_NeurIPS_2019#2]
Hi Harald,

I'm trying to implement this algorithm on a dataset of about 2 million items and 20 million users. And I found that while normalizing XtX a dense mu.T*mu will be needed, which is a dense matrix of shape (2e6, 2e6), making it impossible to calculate even if on spark. Any ideas or suggestions about how to apply the algo for such using case?

Thanks in advance.

**Comment by hasteck** [comment]
Hi,

it is correct that the published algorithm cannot deal with such a large number of items (2 million), as the dense item-item matrix used in the first step does not fit into memory. The dense item-item matrix is only used as a simple way for determining the sparse item-item matrix (by using thresholding). The main part of the algorithm is based on the sparse item-item matrix, which may still be computationally feasible for 2 million items for a sufficiently sparse item-item matrix. So, the question is if there is another way to compute the sparse item-item matrix without first computing the dense item-item matrix. As the sparse matrix is determined by the grouping into sets similar items, this may be possible (approximately) by using meta-data etc, e.g., if the items are grouped into categories. This depends, of course, on the kinds of data that are available in a particular application, and a custom-tailored solution may be required. I hope this helps.

**Comment by cmxcn** [comment]
Not sure if I understand this correctly. For example, if the items can be grouped in categories, we can regard items in one category as the markov blankets of each other?

**Comment by hasteck** [comment]
Yes, this might be the simplest way and possibly a useful approximation. I think you got the idea that I tried to describe. I thought of using meta-data/additional item-features, given that this idea was also used in the SLIM paper (Ning, Karypis), called fSLIM. Depending on the data that you have available, other variants of this idea might work better than the simplest version. As a disclaimer, I have not tried this idea, as I did not face the problem yet that the (dense) item-item matrix did not fit into memory.

**Comment by cmxcn** [comment]
Got it. Thank you!

**Comment by yshenaw** [comment]
Hi, I've implemented a memory-efficient version and we can now run MSD dataset on a laptop. But I haven't test millions of items. I hope it can help you.
https://github.com/yshenaw/ME-MRF-CF

**Comment by hasteck** [comment]
Many thanks for your work!
